### PR TITLE
Runtime patch

### PR DESCRIPTION
**Pull request by m-mayran**
If this PR is for Terraform, I acknowledge that I have:

- [ x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [ x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [ x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/third_party/terraform/tests) (for handwritten resources or update tests).
- [x ] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [ ] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
notebooks: added field `reserved_ip_range` to `google_notebooks_runtime`
```

```release-note:bug
notebooks: fixed a bug where`google_notebooks_runtime` can't be updated
```

### CODE REVIEW

**Comment by shuyama1** [comment]
@m-mayran Looks like the test failed. Let me know if you need help debugging! (failed tests `TestAccContainerCluster_withAddons` and `TestAccContainerCluster_withConfidentialNodes` are unrelated and can be ignored. Sorry for the noise)

**Comment by m-mayran** [comment]
I am not sure if related but it seems that Terraform tries to update a value set automatically `softwareConfig.enable_health_monitoring `

```
error=
  | After applying this test step, the plan was not empty.
  | stdout:
  | 
  | 
  | An execution plan has been generated and is shown below.
  | Resource actions are indicated with the following symbols:
  |   ~ update in-place
  | 
  | Terraform will perform the following actions:
  | 
  |   # google_notebooks_runtime.runtime will be updated in-place
  |   ~ resource "google_notebooks_runtime" "runtime" {
  |         id       = "projects/mam-nooage/locations/us-central1/runtimes/tf-test-notebooks-runtimed2hddnz9w2"
  |         location = "us-central1"
  |         metrics  = []
  |         name     = "tf-test-notebooks-runtimed2hddnz9w2"
  |         project  = "mam-nooage"
  |         state    = "PROVISIONING"
  | 
  |         access_config {
  |             access_type   = "SINGLE_USER"
  |             proxy_uri     = "364d6d1039a5d875-dot-us-central1.notebooks.googleusercontent.com"
  |             runtime_owner = "admin@hashicorptest.com"
  |         }
  | 
  |       ~ software_config {
  |           ~ enable_health_monitoring = false -> true
  |             idle_shutdown            = true
  |             idle_shutdown_timeout    = 80
  |             install_gpu_driver       = false
  |         }
  | 
  |         virtual_machine {
  |             virtual_machine_config {
  |                 guest_attributes = {}
  |                 internal_ip_only = false
  |                 labels           = {}
  |                 machine_type     = "n1-standard-4"
  |                 metadata         = {}
  |                 tags             = []
  |                 zone             = "us-central1-b"
  | 
  |                 data_disk {
  |                     auto_delete       = false
  |                     boot              = false
  |                     guest_os_features = []
  |                     index             = 0
  |                     licenses          = []
  | 
  |                     initialize_params {
  |                         disk_size_gb = 100
  |                         disk_type    = "PD_STANDARD"
  |                         labels       = {}
  |                     }
  |                 }
  |             }
  |         }
  |     }
  | 
  | Plan: 0 to add, 1 to change, 0 to destroy.

```

How can I avoid this in the rules? I do have 

```
softwareConfig: !ruby/object:Overrides::Terraform::PropertyOverride
        default_from_api: true
```

**Comment by m-mayran** [comment]
Also, I see that the state is in PROVISIONING (which might be why enable_health_monitoring is not yet true). Is there something that I can add (and should I) in my tests to wait for the terraform task to be finished?

File: mmv1/products/notebooks/api.yaml
Comment (shuyama1) [inline mmv1/products/notebooks/api.yaml:874]: Any reasons that we want to remove the default value here? I thought the field would always be set to true if the user does not explicitly set this field.
 Plus, removing a default value is likely be considered as a breaking change and we may need to wait until major release to implement it.

File: .gitignore
Comment (shuyama1) [inline .gitignore:23]: Does this need to be added it to the MM repository?

Comment (m-mayran) [reply .gitignore:23]: When I pushed from my environment, it tried to modify the Ruby setup so I though I would add it in the ignore..

File: mmv1/products/notebooks/api.yaml
Comment (shuyama1) [reply mmv1/products/notebooks/api.yaml:874]: Then what's gonna happen if user does not set this field explicitly after removing the default value from Terraform? Will this field be set to true, false, unset or it depends. As we currently set the default value as `True`, which means if user doesn't set this field, Terraform will set it as `True` and send it to the API. If we remove the default value from Terraform now and the API does not set it to `True`, which will possibly break the previous users  and it would be consider as a breaking change.

Comment (m-mayran) [reply mmv1/products/notebooks/api.yaml:874]: What do you suggest then? The team agreed to not change the API behavior and leave it at False. I agree that it is a breaking change. The risk I see is that a Runtime would get deleted and recreated which, unless I am mistaken, is happening today if we don't change this because the default value does not match the API. If we were to go with the breaking change, when would be the next major release? If too far away, is there any way to merge this change beforehand.

Comment (shuyama1) [reply mmv1/products/notebooks/api.yaml:874]: We don't have a clear plan on when the next major release is going to be, but it will likely not happen until next year. I'll have to discuss with my team to see if there's any ways we can work around this problem. 

A quick question, as I'm trying to understand the behavior. Do you know what value will be set to this field if users do not configure it? It should be either true or false, since it is a boolean, right? Looks like the API does not return the value, though.

Comment (m-mayran) [reply mmv1/products/notebooks/api.yaml:874]: Nothing gets set if it is not actively provided. Our control plane will consider that it is false if the key is not there.

If set, it will add an optional attribute `report-system-health` to the underlying GCE instance. Otherwise that attribute won't exist.

If this is acceptable behavior, we would really need to find a solution before next year as we have customers expecting a Terraform update (and at this point, I think having some of their Runtime deleted would be better than not having the feature)

Comment (shuyama1) [reply mmv1/products/notebooks/api.yaml:874]: Okay! Understand! Thanks for your explanation. Will working on finding possible solutions to implement the feature and avoid breaking changes!

Comment (shuyama1) [reply mmv1/products/notebooks/api.yaml:874]: Just talked with the team, I think we can't change the default value until the next major release, but we can at least implement the update feature. I will run a couple of local tests later this afternoon and will leave my suggestion after. Sorry for the delay on this issue.

**Review by shuyama1 — CHANGES_REQUESTED** [review]

File: mmv1/products/notebooks/api.yaml
Comment (shuyama1) [inline mmv1/products/notebooks/api.yaml:151]: Looks like we have `input: true` for this resource which means the resource can not be undated. I wonder if the update functionality works for this resource after adding `PATCH`. Can we also add a test?

File: mmv1/products/notebooks/api.yaml
Comment (shuyama1) [reply mmv1/products/notebooks/api.yaml:871]: Just tested locally. Adding 
```
update_mask_fields:
          - "softwareConfig.idle_shutdown"
          - "softwareConfig.idle_shutdown_timeout"
          - "softwareConfig.custom_gpu_driver_path"
          - "softwareConfig.post_startup_script"
 ```
to https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/products/notebooks/terraform.yaml#L164 with `update_verb: :PATCH` and `update_mask: true` set on the resource level should be working.

File: mmv1/products/notebooks/api.yaml
Comment (m-mayran) [reply mmv1/products/notebooks/api.yaml:821]: This requires to setup VPC Peering in the test project which adds quite a lot of complexity. As discussed, we will skip this for now.

File: mmv1/products/notebooks/api.yaml
Comment (m-mayran) [reply mmv1/products/notebooks/api.yaml:151]: Why did we add those lines? This PR is supposed to be for `Runtime` not `Instance`.

Comment (shuyama1) [reply mmv1/products/notebooks/api.yaml:151]: I agree. We can exclude these two lines from this PR. I just saw these two lines added in the PR and if that is intentionally, I would suggest that we add an update test for it.

File: .gitignore
Comment (m-mayran) [reply .gitignore]: I removed the file

File: mmv1/products/notebooks/api.yaml
Comment (m-mayran) [reply mmv1/products/notebooks/api.yaml]: Done.

**Comment by shuyama1** [comment]
@m-mayran Sorry that I forgot to mention in the last review. We probably want to add `input: true` to all the fields that are immutable, as it will tell Terraform to delete and recreate the resource when users try to change the values of those fields.

**Comment by m-mayran** [comment]
> @m-mayran Looks like the test failed. Let me know if you need help debugging! (failed tests `TestAccContainerCluster_withAddons` and `TestAccContainerCluster_withConfidentialNodes` are unrelated and can be ignored. Sorry for the noise)

We might change some of the fields that can be edited over time. Is it ok to leave those without the `input: true` for now?

Also, if I look the Runtime's name for example, that one is immutable and has `output: true`. Would it still need `input: true`?

**Comment by shuyama1** [comment]
> > @m-mayran Looks like the test failed. Let me know if you need help debugging! (failed tests `TestAccContainerCluster_withAddons` and `TestAccContainerCluster_withConfidentialNodes` are unrelated and can be ignored. Sorry for the noise)
> 
> We might change some of the fields that can be edited over time. Is it ok to leave those without the `input: true` for now?
> 
> Also, if I look the Runtime's name for example, that one is immutable and has `output: true`. Would it still need `input: true`?

Output-only fields don't need them. I would recommend you at least add it to fields whose update functionality can't be supported for sure or can't be supported in the near future. Plus, removing `input: true` does not consider a breaking change. We can remove it from the field anytime when the update logic is supported.

**Comment by m-mayran** [comment]
> > > @m-mayran Looks like the test failed. Let me know if you need help debugging! (failed tests `TestAccContainerCluster_withAddons` and `TestAccContainerCluster_withConfidentialNodes` are unrelated and can be ignored. Sorry for the noise)
> > 
> > 
> > We might change some of the fields that can be edited over time. Is it ok to leave those without the `input: true` for now?
> > Also, if I look the Runtime's name for example, that one is immutable and has `output: true`. Would it still need `input: true`?
> 
> Output-only fields don't need them. I would recommend you at least add it to fields whose update functionality can't be supported for sure or can't be supported in the near future. Plus, removing `input: true` does not consider a breaking change. We can remove it from the field anytime when the update logic is supported.

Thanks for the details. Added to the fields that will most likely be immutable until we set up an update for GCE fields who don't required an instance restart.

File: mmv1/products/notebooks/api.yaml
Comment (shuyama1) [inline mmv1/products/notebooks/api.yaml:829]:
```suggestion
                  Reserved IP Range name is used for VPC Peering. The
```
We can remove `Optional` from the description, as it would auto generated `(optional)` for the field in the downstream documentation if we don't set the field `required` in the yaml file.

**Review by shuyama1 — CHANGES_REQUESTED** [review]
LGTM in general! It should be good to go after adding the test for `reserved_ip_range`. Thank you so much for the work!

**Review by shuyama1 — APPROVED** [review]
Thank you!

**Pull request merged**
